File: src/main/java/io/outfox/typescriptpoet/CodeWriter.kt
/*
 * Copyright 2017 Outfox, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.outfox.typescriptpoet

import java.util.*


/**
 * Converts a [FileSpec] to a string suitable to both human- and tsc-consumption. This honors
 * imports, indentation, and deferred variable names.
 */
internal class CodeWriter constructor(
   out: Appendable,
   private val indent: String = "  ",
   referencedSymbols: Set<SymbolSpec> = emptySet()
) : SymbolReferenceTracker {
  private val out = LineWrapper(out, indent, 100)
  private var indentLevel = 0

  private var javaDoc = false
  private var comment = false
  private var trailingNewline = false

  private var referencedSymbols = referencedSymbols.toMutableSet()

  override fun referenced(symbol: SymbolSpec) {
    referencedSymbols.add(symbol)
  }

  /**
   * When emitting a statement, this is the line of the statement currently being written. The first
   * line of a statement is indented normally and subsequent wrapped lines are double-indented. This
   * is -1 when the currently-written line isn't part of a statement.
   */
  private var statementLine = -1

  fun indent(levels: Int = 1) = apply {
    indentLevel += levels
  }

  fun unindent(levels: Int = 1) = apply {
    require(indentLevel - levels >= 0) { "cannot unindent $levels from $indentLevel" }
    indentLevel -= levels
  }

  fun emitComment(codeBlock: CodeBlock) {
    trailingNewline = true // Force the '//' prefix for the comment.
    comment = true
    try {
      emitCode(codeBlock)
      emit("\n")
    }
    finally {
      comment = false
    }
  }

  fun emitJavaDoc(javaDocCodeBlock: CodeBlock) {
    if (javaDocCodeBlock.isEmpty()) return

    emit("/**\n")
    javaDoc = true
    try {
      emitCode(javaDocCodeBlock)
    }
    finally {
      javaDoc = false
    }
    emit(" */\n")
  }

  fun emitDecorators(decorators: List<DecoratorSpec>, inline: Boolean) {
    for (decoratorSpec in decorators) {
      decoratorSpec.emit(this, inline)
      emit(if (inline) " " else "\n")
    }
  }

  /**
   * Emits `modifiers` in the standard order. Modifiers in `implicitModifiers` will not
   * be emitted.
   */
  fun emitModifiers(
     modifiers: Set<Modifier>,
     implicitModifiers: Set<Modifier> = emptySet()) {
    if (modifiers.isEmpty()) return
    for (modifier in EnumSet.copyOf(modifiers)) {
      if (implicitModifiers.contains(modifier)) continue
      emit(modifier.keyword)
      emit(" ")
    }
  }

  /**
   * Emit type variables with their bounds.
   *
   * This should only be used when declaring type variables; everywhere else bounds are omitted.
   */
  fun emitTypeVariables(typeVariables: List<TypeName.TypeVariable>) {
    if (typeVariables.isEmpty()) return

    emit("<")
    typeVariables.forEachIndexed { index, typeVariable ->
      if (index > 0) emit(", ")
      emitCode(buildString {
        append(typeVariable.name)
        if (typeVariable.bounds.isNotEmpty()) {
          val parts = mutableListOf<String>()
          parts.add(" extends")
          typeVariable.bounds.forEachIndexed { index, bound ->
            if (index > 0) parts.add(bound.combiner.symbol)
            bound.modifier?.let { parts.add(it.keyword) }
            parts.add(bound.type.reference(this@CodeWriter))
          }
          append(parts.joinToString(" "))
        }
      })
    }
    emit(">")
  }

  fun emitCode(s: String) = emitCode(CodeBlock.of(s))

  fun emitCode(format: String, vararg args: Any?) = emitCode(CodeBlock.of(format, *args))

  fun emitCode(codeBlock: CodeBlock) = apply {

    // Transfer all symbols referenced in the code block
    this@CodeWriter.referencedSymbols.addAll(codeBlock.referencedSymbols)

    var a = 0
    val partIterator = codeBlock.formatParts.listIterator()
    while (partIterator.hasNext()) {
      val part = partIterator.next()
      when (part) {
        "%L" -> emitLiteral(codeBlock.args[a++])

        "%N" -> emit(codeBlock.args[a++] as String)

        "%S" -> emitString(codeBlock.args[a++] as String?)

        "%T" -> emitTypeName(codeBlock.args[a++] as TypeName)

        "%%" -> emit("%")

        "%>" -> indent()

        "%<" -> unindent()

        "%[" -> beginStatement()

        "%]" -> endStatement()

        "%W" -> emitWrappingSpace()

        else -> {
          // Handle deferred type.
          emit(part)
        }
      }
    }
  }

  private fun beginStatement() {
    check(statementLine == -1) { "statement enter %[ followed by statement enter %[" }

    statementLine = 0
  }

  private fun endStatement() {
    check(statementLine != -1) { "statement exit %] has no matching statement enter %[" }

    if (statementLine > 0) {
      unindent(2) // End a multi-line statement. Decrease the indentation level.
    }

    statementLine = -1
  }

  private fun emitWrappingSpace() = apply {
    out.wrappingSpace(indentLevel + 2)
  }

  private fun emitTypeName(typeName: TypeName) {
    emit(typeName.reference(this))
  }

  private fun emitString(string: String?) {
    // Emit null as a literal null: no quotes.
    emit(if (string != null)
           stringLiteralWithQuotes(string)
         else
           "null")
  }

  private fun emitLiteral(o: Any?) {
    when (o) {
      is ClassSpec -> o.emit(this)
      is InterfaceSpec -> o.emit(this)
      is EnumSpec -> o.emit(this)
      is DecoratorSpec -> o.emit(this, inline = true, asParameter = true)
      is CodeBlock -> emitCode(o)
      else -> emit(o.toString())
    }
  }

  /**
   * Emits `s` with indentation as required. It's important that all code that writes to
   * [CodeWriter.out] does it through here, since we emit indentation lazily in order to avoid
   * unnecessary trailing whitespace.
   */
  fun emit(s: String) = apply {
    var first = true
    for (line in s.split('\n')) {
      // Emit a newline character. Make sure blank lines in KDoc & comments look good.
      if (!first) {
        if ((javaDoc || comment) && trailingNewline) {
          emitIndentation()
          out.append(if (javaDoc) " *" else "//")
        }
        out.append("\n")
        trailingNewline = true
        if (statementLine != -1) {
          if (statementLine == 0) {
            indent(2) // Begin multiple-line statement. Increase the indentation level.
          }
          statementLine++
        }
      }

      first = false
      if (line.isEmpty()) continue // Don't indent empty lines.

      // Emit indentation and comment prefix if necessary.
      if (trailingNewline) {
        emitIndentation()
        if (javaDoc) {
          out.append(" * ")
        }
        else if (comment) {
          out.append("// ")
        }
      }

      out.append(line)
      trailingNewline = false
    }
  }

  private fun emitIndentation() {
    for (j in 0 until indentLevel) {
      out.append(indent)
    }
  }

  /**
   * Returns the symbols that are required to be imported for this code. If there were any simple name
   * collisions, that symbol's first use is imported; which may cause compilation issues.
   */
  fun requiredImports(): Set<SymbolSpec.Imported> {
    return referencedSymbols.filterIsInstance<SymbolSpec.Imported>().toImmutableSet()
  }
}

File: src/CodeWriter.ts
import _ from 'lodash';
import * as Path from 'path';
import { ClassSpec } from './ClassSpec';
import { CodeBlock } from './CodeBlock';
import { DecoratorSpec } from './DecoratorSpec';
import { EnumSpec } from './EnumSpec';
import { FunctionSpec } from './FunctionSpec';
import { InterfaceSpec } from './InterfaceSpec';
import { LineWrapper } from './LineWrapper';
import { Modifier, ModifierOrder } from './Modifier';
import { StringBuffer } from './StringBuffer';
import { SymbolReferenceTracker } from './SymbolReferenceTracker';
import { Augmented, Imported, ImportsAll, ImportsDefault, ImportsName, SideEffect, SymbolSpec } from './SymbolSpecs';
import { TypeName, TypeVariable } from './TypeNames';
import { check, filterInstances, stringLiteralWithQuotes, unique } from './utils';

/**
 * Converts a [FileSpec] to a string suitable to both human- and tsc-consumption. This honors
 * imports, indentation, and deferred variable names.
 */
export class CodeWriter implements SymbolReferenceTracker {
  public static emitToString(emittable: { emit(cw: CodeWriter): void }): string {
    const out = new StringBuffer();
    emittable.emit(new CodeWriter(out));
    return out.toString();
  }

  private readonly out: LineWrapper;
  private readonly referencedSymbols: Set<SymbolSpec> = new Set();
  private indentLevel = 0;
  private javaDoc = false;
  private comment = false;
  private trailingNewline = false;

  constructor(out: StringBuffer, private indentString: string = '  ', referencedSymbols: Set<SymbolSpec> = new Set()) {
    this.out = new LineWrapper(out, indentString, 100);
    referencedSymbols.forEach(sym => this.referencedSymbols.add(sym));
  }

  public referenced(symbol: SymbolSpec): void {
    this.referencedSymbols.add(symbol);
  }

  public indent(levels: number = 1): this {
    this.indentLevel += levels;
    return this;
  }

  public unindent(levels: number = 1): this {
    check(this.indentLevel - levels >= 0, `cannot unindent ${levels} from ${this.indentLevel}`);
    this.indentLevel -= levels;
    return this;
  }

  public emitComment(codeBlock: CodeBlock): void {
    this.trailingNewline = true; // Force the '//' prefix for the comment.
    this.comment = true;
    try {
      this.emitCodeBlock(codeBlock);
      this.emit('\n');
    } finally {
      this.comment = false;
    }
  }

  public emitJavaDoc(javaDocCodeBlock: CodeBlock): void {
    if (javaDocCodeBlock.isEmpty()) {
      return;
    }
    this.emit('/**\n');
    this.javaDoc = true;
    try {
      this.emitCodeBlock(javaDocCodeBlock);
    } finally {
      this.javaDoc = false;
    }
    this.emit(' */\n');
  }

  public emitDecorators(decorators: DecoratorSpec[], inline: boolean): void {
    decorators.forEach(decoratorSpec => {
      decoratorSpec.emit(this, inline);
      this.emit(inline ? ' ' : '\n');
    });
  }

  /**
   * Emits `modifiers` in the standard order. Modifiers in `implicitModifiers` will not
   * be emitted.
   */
  public emitModifiers(modifiers: Modifier[], implicitModifiers: Modifier[] = []): void {
    if (modifiers.length === 0) {
      return;
    }
    ModifierOrder.forEach(m => {
      if (modifiers.includes(m) && !implicitModifiers.includes(m)) {
        this.emit(m).emit(' ');
      }
    });
  }

  /**
   * Emit type variables with their bounds.
   *
   * This should only be used when declaring type variables; everywhere else bounds are omitted.
   */
  public emitTypeVariables(typeVariables: TypeVariable[]): void {
    if (typeVariables.length === 0) {
      return;
    }
    this.emit('<');
    let index = 0;
    typeVariables.forEach(typeVariable => {
      if (index > 0) {
        this.emit(', ');
      }
      let code = typeVariable.name;
      if (typeVariable.bounds.length > 0) {
        const parts: string[] = [];
        parts.push(' extends');
        let j = 0;
        typeVariable.bounds.forEach(bound => {
          if (j > 0) {
            parts.push(bound.combiner);
          }
          if (bound.modifier) {
            parts.push(bound.modifier);
          }
          parts.push(bound.type.reference(this));
          j++;
        });
        code += parts.join(' ');
      }
      this.emit(code);
      index++;
    });
    this.emit('>');
  }

  public emitImports(path: string): this {
    const imports = this.requiredImports();
    const augmentImports = _.groupBy(filterInstances(imports, Augmented), a => a.augmented);
    const sideEffectImports = _.groupBy(filterInstances(imports, SideEffect), a => a.source);

    if (imports.length > 0) {
      const m = _.groupBy(
        imports.filter(it => !(it instanceof Augmented) || !(it instanceof SideEffect)),
        it => it.source
      ); // FileModules.importPath(this.path, it.source));
      // .toSortedMap()
      // tslint:disable-next-line:no-shadowed-variable
      Object.entries(m).forEach(([sourceImportPath, imports]) => {
        // Skip imports from the current module
        if (path === sourceImportPath || Path.resolve(path) === Path.resolve(sourceImportPath)) {
          return;
        }
        // Output star imports individually
        filterInstances(imports, ImportsAll).forEach(i => {
          this.emitCode("%[import * as %L from '%L';\n%]", i.value, sourceImportPath);
          const augments = augmentImports[i.value];
          if (augments) {
            augments.forEach(augment => this.emitCode("%[import '%L';\n%]", augment.source));
          }
        });
        // Output named imports as a group
        const names = unique(filterInstances(imports, ImportsName).map(it => it.value));
        const def = unique(filterInstances(imports, ImportsDefault).map(it => it.value));
        if (names.length > 0 || def.length > 0) {
          const namesPart = names.length > 0 ? [`{ ${names.join(', ')} }`] : [];
          const defPart = def.length > 0 ? [def[0]] : [];
          this.emitCode('import ')
            .emitCode([...defPart, ...namesPart].join(', '))
            .emitCode(" from '%L';\n", sourceImportPath);
          [...names, ...def].forEach(name => {
            const augments = augmentImports[name];
            if (augments) {
              augments.forEach(augment => this.emitCode("%[import '%L';\n%]", augment.source));
            }
          });
        }
      });

      Object.keys(sideEffectImports).forEach(it => {
        this.emitCode('%[import %S;\n%]', it);
      });

      this.emit('\n');
    }
    return this;
  }

  /* TODO
  public emitCode(s: string): void {
    this.emitCodeBlock(CodeBlock.of(s));
  }
  */

  public emitCode(format: string, ...args: unknown[]): this {
    this.emitCodeBlock(CodeBlock.of(format, ...args));
    return this;
  }

  public emitCodeBlock(codeBlock: CodeBlock): this {
    // Transfer all symbols referenced in the code block
    codeBlock.referencedSymbols.forEach(sym => this.referencedSymbols.add(sym));
    let a = 0;
    codeBlock.formatParts.forEach(part => {
      switch (part) {
        case '%L':
          this.emitLiteral(codeBlock.args[a++]);
          break;
        case '%N':
          this.emit(codeBlock.args[a++] as string);
          break;
        case '%S':
          this.emitString(codeBlock.args[a++] as string);
          break;
        case '%T':
          this.emitTypeName(codeBlock.args[a++] as TypeName);
          break;
        case '%%':
          this.emit('%');
          break;
        case '%>':
          this.indent();
          break;
        case '%<':
          this.unindent();
          break;
        case '%[':
          break;
        case '%]':
          break;
        case '%W':
          this.emitWrappingSpace();
          break;
        case '%F':
          this.emitFunction(codeBlock.args[a++] as FunctionSpec);
          break;
        // Handle deferred type.
        default:
          this.emit(part);
      }
    });
    if (codeBlock.trailer) {
      this.emitCodeBlock(codeBlock.trailer);
    }
    return this;
  }
  /**
   * Emits `s` with indentation as required. It's important that all code that writes to
   * [CodeWriter.out] does it through here, since we emit indentation lazily in order to avoid
   * unnecessary trailing whitespace.
   */
  public emit(s: string): this {
    let first = true;
    s.split('\n').forEach(line => {
      // Emit a newline character. Make sure blank lines in KDoc & comments look good.
      if (!first) {
        if ((this.javaDoc || this.comment) && this.trailingNewline) {
          this.emitIndentation();
          this.out.append(this.javaDoc ? ' *' : '//');
        }
        this.out.append('\n');
        this.trailingNewline = true;
      }

      first = false;
      if (line.length === 0) {
        return; // Don't indent empty lines.
      }

      // Emit indentation and comment prefix if necessary.
      if (this.trailingNewline) {
        this.emitIndentation();
        if (this.javaDoc) {
          this.out.append(' * ');
        } else if (this.comment) {
          this.out.append('// ');
        }
      }

      this.out.append(line);
      this.trailingNewline = false;
    });

    return this;
  }

  public newLine(): this {
    return this.emit('\n');
  }

  /**
   * Returns the symbols that are required to be imported for this code. If there were any simple name
   * collisions, that symbol's first use is imported; which may cause compilation issues.
   */
  public requiredImports(): Imported[] {
    const imported: Imported[] = [];
    this.referencedSymbols.forEach(sym => {
      if (sym instanceof Imported) {
        imported.push(sym);
      }
    });
    return imported;
  }

  private emitIndentation(): void {
    for (let j = 0; j < this.indentLevel; j++) {
      this.out.append(this.indentString);
    }
  }

  private emitWrappingSpace(): this {
    this.out.wrappingSpace(this.indentLevel + 2);
    return this;
  }

  private emitTypeName(typeName: TypeName): void {
    this.emit(typeName.reference(this));
  }

  private emitFunction(fn: FunctionSpec): void {
    fn.emit(this);
  }

  private emitString(s?: string): void {
    if (s === null) {
      // Emit null as a literal null: no quotes.
      this.emit('null');
    } else if (s === undefined) {
      this.emit('undefined');
    } else {
      this.emit(stringLiteralWithQuotes(s));
    }
  }

  private emitLiteral(o?: unknown): void {
    if (o instanceof ClassSpec) {
      return o.emit(this);
    } else if (o instanceof InterfaceSpec) {
      return o.emit(this);
    } else if (o instanceof EnumSpec) {
      return o.emit(this);
    } else if (o instanceof DecoratorSpec) {
      return o.emit(this, true, true);
    } else if (o instanceof CodeBlock) {
      this.emitCodeBlock(o);
    } else if (typeof o === 'object' && o) {
      this.emit(o.toString());
    } else {
      this.emit(String(o));
    }
  }
}
